File: scripts/multivariate_analysis.py
#!/usr/bin/env python3
"""
Multivariable Regression Analysis (Python)
Supports: linear, logistic, Cox proportional hazards
Medical Statistics Skill
"""
import argparse
import sys
import pandas as pd
import numpy as np
import scipy.stats as stats


def read_data(filepath):
    ext = filepath.split(".")[-1].lower()
    if ext == "csv":
        return pd.read_csv(filepath)
    elif ext in ("xls", "xlsx"):
        return pd.read_excel(filepath)
    elif ext == "sav":
        import pyreadstat
        return pd.read_spss(filepath)
    elif ext == "dta":
        return pd.read_stata(filepath)
    else:
        raise ValueError(f"Unsupported format: {ext}")


def run_linear(df, outcome, predictors):
    import statsmodels.api as sm
    X = sm.add_constant(df[predictors].copy())
    mask = X.notna().all(axis=1) & df[outcome].notna()
    X, y = X[mask], df.loc[mask, outcome]
    model = sm.OLS(y, X).fit()
    print(f"\n  R² = {model.rsquared:.4f}, Adj. R² = {model.rsquared_adj:.4f}")
    print(f"  F = {model.fvalue:.4f}, P = {model.f_pvalue:.4e}\n")
    print("  ┌────────────────────────────────────────────────────┐")
    print(f"  │ {'Variable':<20} │ {'β (95% CI)':<28} │ {'P':<8} │")
    print("  ├────────────────────────────────────────────────────┤")
    rows = []
    for var, coef in model.params.items():
        ci = model.conf_int().loc[var]
        p = model.pvalues[var]
        sig = "***" if p < 0.001 else "**" if p < 0.01 else "*" if p < 0.05 else ""
        print(f"  │ {var:<20} │ {coef:.4f} ({ci[0]:.4f}, {ci[1]:.4f}) │ {p:.4f}{sig:<3} │")
        rows.append({"Variable": var, "Beta": round(coef, 4), "CI_Lower": round(ci[0], 4),
                      "CI_Upper": round(ci[1], 4), "P_value": round(p, 4), "Sig": sig})
    print("  └────────────────────────────────────────────────────┘")
    return model, pd.DataFrame(rows)


def run_logistic(df, outcome, predictors):
    import statsmodels.api as sm
    df[outcome] = pd.Categorical(df[outcome])
    X = sm.add_constant(df[predictors].copy())
    mask = X.notna().all(axis=1) & df[outcome].notna()
    X, y = X[mask], df.loc[mask, outcome]
    model = sm.Logit(y, X).fit(disp=0)
    or_val = np.exp(model.params)
    or_ci = np.exp(model.conf_int())

    print(f"\n  Pseudo R² = {model.prsquared:.4f}")
    print(f"  Log-Likelihood = {model.llf:.2f}")
    print(f"  AUC = {model.bic:.4f} (approximate)")
    print("\n  ┌────────────────────────────────────────────────────────────┐")
    print(f"  │ {'Variable':<20} │ {'OR (95% CI)':<30} │ {'P':<8} │")
    print("  ├────────────────────────────────────────────────────────────┤")
    rows = []
    for var in model.params.index:
        p = model.pvalues[var]
        sig = "***" if p < 0.001 else "**" if p < 0.01 else "*" if p < 0.05 else ""
        print(f"  │ {var:<20} │ {or_val[var]:.4f} ({or_ci[0][var]:.4f}, {or_ci[1][var]:.4f}) │ {p:.4f}{sig:<3} │")
        rows.append({"Variable": var, "OR": round(or_val[var], 4),
                      "CI_Lower": round(or_ci[0][var], 4), "CI_Upper": round(or_ci[1][var], 4),
                      "P_value": round(p, 4), "Sig": sig})
    print("  └────────────────────────────────────────────────────────────┘")
    return model, pd.DataFrame(rows)


def run_cox(df, outcome, predictors, time_col):
    from lifelines import CoxPHFitter
    model_df = df[[time_col, outcome] + predictors].dropna()
    cph = CoxPHFitter()
    cph.fit(model_df, duration_col=time_col, event_col=outcome)
    hr = np.exp(cph.params_)
    hr_ci = np.exp(cph.confidence_intervals_)

    print(f"\n  Concordance = {cph.concordance_index_:.4f}")
    print(f"  Log-Likelihood = {cph.log_likelihood_:.2f}")
    print(f"  AIC = {cph.AIC_partial_:.2f}")
    print("\n  ┌────────────────────────────────────────────────────────────┐")
    print(f"  │ {'Variable':<20} │ {'HR (95% CI)':<30} │ {'P':<8} │")
    print("  ├────────────────────────────────────────────────────────────┤")
    rows = []
    for var in cph.params_.index:
        p = cph.summary.loc[var, "p"]
        sig = "***" if p < 0.001 else "**" if p < 0.01 else "*" if p < 0.05 else ""
        print(f"  │ {var:<20} │ {hr[var]:.4f} ({hr_ci[var].iloc[0]:.4f}, {hr_ci[var].iloc[1]:.4f}) │ {p:.4f}{sig:<3} │")
        rows.append({"Variable": var, "HR": round(hr[var], 4),
                      "CI_Lower": round(hr_ci[var].iloc[0], 4), "CI_Upper": round(hr_ci[var].iloc[1], 4),
                      "P_value": round(p, 4), "Sig": sig})
    print("  └────────────────────────────────────────────────────────────┘")
    return cph, pd.DataFrame(rows)


def main():
    parser = argparse.ArgumentParser(description="Multivariable Regression")
    parser.add_argument("--data", required=True, help="Data file path")
    parser.add_argument("--outcome", required=True, help="Outcome variable")
    parser.add_argument("--type", required=True, choices=["linear", "logistic", "cox"], help="Model type")
    parser.add_argument("--vars", help="Comma-separated predictor variables")
    parser.add_argument("--time", help="Time variable (required for Cox)")
    parser.add_argument("--pthreshold", type=float, default=0.1, help="Variable selection threshold")
    args = parser.parse_args()

    df = read_data(args.data)

    # Determine predictors
    if args.vars:
        predictors = [v.strip() for v in args.vars.split(",") if v.strip() in df]
    else:
        if args.type == "cox" and args.time:
            predictors = [c for c in df.columns if c not in (args.outcome, args.time)]
        else:
            predictors = [c for c in df.columns if c != args.outcome]

    print("\n" + "=" * 65)
    model_names = {"linear": "Linear", "logistic": "Logistic", "cox": "Cox PH"}
    print(f"         {model_names[args.type]} Regression Results")
    print("=" * 65)
    print(f"\n  N = {len(df)}")
    print(f"  Predictors: {', '.join(predictors)}\n")

    if args.type == "linear":
        model, result_df = run_linear(df, args.outcome, predictors)
        export_prefix = "linear"
    elif args.type == "logistic":
        model, result_df = run_logistic(df, args.outcome, predictors)
        export_prefix = "logistic"
    elif args.type == "cox":
        if not args.time:
            print("ERROR: --time required for Cox regression")
            sys.exit(1)
        model, result_df = run_cox(df, args.outcome, predictors, args.time)
        export_prefix = "cox"

    # Export results
    from export_utils import export_to_excel, export_to_word, timestamp
    ts = timestamp()
    export_to_excel(result_df, f"{export_prefix}_{ts}.xlsx",
                    sheet_name=export_prefix,
                    title=f"{model_names[args.type]} Regression Results")
    export_to_word(result_df, f"{export_prefix}_{ts}.docx",
                   title=f"{model_names[args.type]} Regression Results")

    # Forest plot
    try:
        from forest_plot import forest_plot as fp
        effect_map = {"linear": ("beta (95% CI)", 0),
                      "logistic": ("OR (95% CI)", 1),
                      "cox": ("HR (95% CI)", 1)}
        eff_label, ref_val = effect_map.get(args.type, ("OR (95% CI)", 1))

        # Filter out intercept for linear/logistic
        plot_df = result_df.copy()
        if "Intercept" in plot_df["Variable"].values:
            plot_df = plot_df[plot_df["Variable"] != "Intercept"]

        if len(plot_df) > 0:
            fp(plot_df, effect_label=eff_label, reference=ref_val,
               filename_prefix=f"forest_{export_prefix}_{ts}",
               title=f"Forest Plot: {model_names[args.type]}")
    except Exception as e:
        print(f"  ! Forest plot skipped: {e}")

    print("\n" + "=" * 65)


if __name__ == "__main__":
    main()

File: scripts/multivariate_analysis.R
#!/usr/bin/env Rscript
# ============================================
# 多因素回归分析脚本
# medical-statistics skill
# ============================================

suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(haven))
suppressPackageStartupMessages(library(readxl))
suppressPackageStartupMessages(library(survival))

args <- commandArgs(trailingOnly = TRUE)

parse_args <- function(args) {
  params <- list()
  i <- 1
  while (i <= length(args)) {
    if (args[i] == "--data" && i + 1 <= length(args)) {
      params$data <- args[i + 1]
      i <- i + 2
    } else if (args[i] == "--outcome" && i + 1 <= length(args)) {
      params$outcome <- args[i + 1]
      i <- i + 2
    } else if (args[i] == "--type" && i + 1 <= length(args)) {
      params$type <- args[i + 1]
      i <- i + 2
    } else if (args[i] == "--vars" && i + 1 <= length(args)) {
      params$vars <- unlist(strsplit(args[i + 1], ","))
      i <- i + 2
    } else if (args[i] == "--time" && i + 1 <= length(args)) {
      params$time <- args[i + 1]
      i <- i + 2
    } else if (args[i] == "--pthreshold" && i + 1 <= length(args)) {
      params$pthreshold <- as.numeric(args[i + 1])
      i <- i + 2
    } else {
      i <- i + 1
    }
  }
  return(params)
}

params <- parse_args(args)

if (is.null(params$data) || is.null(params$outcome) || is.null(params$type)) {
  stop("请指定: --data <path> --outcome <variable> --type <linear|logistic|cox>")
}

if (is.null(params$pthreshold)) params$pthreshold <- 0.1

# 读取数据
ext <- tolower(tools::file_ext(params$data))
if (ext == "csv") {
  df <- read.csv(params$data, stringsAsFactors = FALSE)
} else if (ext %in% c("xls", "xlsx")) {
  df <- read_excel(params$data)
} else if (ext %in% c("sav")) {
  df <- read_sav(params$data)
} else if (ext %in% c("dta")) {
  df <- read_dta(params$data)
} else {
  stop("不支持的文件格式: ", ext)
}

# 构建公式
if (is.null(params$vars)) {
  # 所有其他变量作为自变量
  if (params$type == "cox") {
    if (is.null(params$time)) stop("Cox回归需要指定 --time")
    vars <- setdiff(names(df), c(params$outcome, params$time))
  } else {
    vars <- setdiff(names(df), params$outcome)
  }
} else {
  vars <- params$vars[vars %in% names(df)]
}

cat("\n═══════════════════════════════════════════\n")
cat(sprintf("     多因素分析: %s回归\n", switch(params$type,
  "linear" = "线性",
  "logistic" = "Logistic",
  "cox" = "Cox比例风险"
)))
cat("═══════════════════════════════════════════\n\n")

cat("纳入变量:", paste(vars, collapse = ", "), "\n\n")

# 执行回归
if (params$type == "linear") {
  formula <- as.formula(paste(params$outcome, "~", paste(vars, collapse = " + ")))
  model <- lm(formula, data = df)
  s <- summary(model)

  cat("━━━ 模型摘要 ━━━\n")
  cat(sprintf("R² = %.4f, 调整R² = %.4f\n", s$r.squared, s$adj.r.squared))
  cat(sprintf("F = %.4f, P = %.4e\n\n", s$fstatistic[1], pf(s$fstatistic[1], s$fstatistic[2], s$fstatistic[3], lower.tail = FALSE)))

  coef_df <- as.data.frame(coef(s))
  names(coef_df) <- c("β", "Std.Error", "t_value", "P_value")

  cat("━━━ 回归系数 ━━━\n")
  result <- data.frame(
    变量 = rownames(coef_df),
    β = round(coef_df$β, 4),
    SE = round(coef_df$Std.Error, 4),
    t = round(coef_df$t_value, 3),
    P = round(coef_df$P_value, 4),
    `95%CI_L` = round(coef_df$β - 1.96 * coef_df$Std.Error, 4),
    `95%CI_U` = round(coef_df$β + 1.96 * coef_df$Std.Error, 4),
    stringsAsFactors = FALSE
  )
  result$显著性 <- ifelse(result$P < 0.001, "***",
                   ifelse(result$P < 0.01, "**",
                   ifelse(result$P < 0.05, "*", "")))
  print(result, row.names = FALSE)

} else if (params$type == "logistic") {
  df[[params$outcome]] <- as.factor(df[[params$outcome]])
  formula <- as.formula(paste(params$outcome, "~", paste(vars, collapse = " + ")))
  model <- glm(formula, data = df, family = binomial())
  s <- summary(model)

  or <- exp(coef(model))
  ci <- exp(confint(model))

  cat("━━━ Logistic回归结果 (OR) ━━━\n\n")
  result <- data.frame(
    变量 = names(coef(model)),
    OR = round(or, 4),
    `95%CI_L` = round(ci[, 1], 4),
    `95%CI_U` = round(ci[, 2], 4),
    P = round(coef(s)[, 4], 4),
    stringsAsFactors = FALSE
  )
  result$OR_CI <- sprintf("%.3f (%.3f-%.3f)", result$OR, result$X95CI_L, result$X95CI_U)
  result$显著性 <- ifelse(result$P < 0.001, "***",
                   ifelse(result$P < 0.01, "**",
                   ifelse(result$P < 0.05, "*", "")))
  print(result[, c("变量", "OR_CI", "P", "显著性")], row.names = FALSE)

} else if (params$type == "cox") {
  formula <- as.formula(paste0("Surv(", params$time, ", ", params$outcome, ") ~ ", paste(vars, collapse = " + ")))
  model <- coxph(formula, data = df)
  s <- summary(model)

  hr <- exp(coef(model))
  ci <- exp(confint(model))

  cat("━━━ Cox回归结果 (HR) ━━━\n\n")
  result <- data.frame(
    变量 = names(hr),
    HR = round(hr, 4),
    `95%CI_L` = round(ci[, 1], 4),
    `95%CI_U` = round(ci[, 2], 4),
    P = round(s$coefficients[, "Pr(>|z|)"], 4),
    stringsAsFactors = FALSE
  )
  result$HR_CI <- sprintf("%.3f (%.3f-%.3f)", result$HR, result$X95CI_L, result$X95CI_U)
  result$显著性 <- ifelse(result$P < 0.001, "***",
                   ifelse(result$P < 0.01, "**",
                   ifelse(result$P < 0.05, "*", "")))
  print(result[, c("变量", "HR_CI", "P", "显著性")], row.names = FALSE)
}

cat("\n\n━━━━━━━━━━━━━━━━━━ 模型信息 ━━━━━━━━━━━━━━━━━━\n")
cat(sprintf("回归类型: %s\n", params$type))
cat(sprintf("样本量: %d\n", nrow(model$model)))
cat(sprintf("变量数: %d\n", length(vars)))

cat("\n═══════════════════════════════════════════\n")
